include/gatekeeper_l2.h

```diff
@@ -0,0 +1,116 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_L2_H_
+#define _GATEKEEPER_L2_H_
+
+/*
+ * Return the L2 header length of a received packet.
+ *
+ * WARNING
+ *	Note that in order to use this function, @pkt must have first gone
+ *	through pkt_in_skip_l2() or another function to set its packet type.
+ */
+static inline size_t
+pkt_in_l2_hdr_len(struct rte_mbuf *pkt)
+{
+	size_t l2_len = sizeof(struct ether_hdr);
+
+	switch (pkt->l2_type) {
+	case RTE_PTYPE_L2_ETHER_VLAN:
+		l2_len += sizeof(struct vlan_hdr);
+		break;
+	default:
+		break;
+	}
+
+	return l2_len;
+}
+
+/*
+ * Return the difference in the size of the L2 header between
+ * packet received (@pkt) and a packet that will be transmitted
+ * on @iface. This determines what changes have to be made to the
+ * L2 space of the packet.
+ *
+ * A negative number indicates that bytes need to be removed from
+ * the L2 space, and a positive number indicates that bytes need to
+ * be added to the L2 space.
+ *
+ * WARNING
+ *	Note that in order to use this function, @pkt must have first gone
+ *	through pkt_in_skip_l2() or another function to set its packet type.
+ */
+static inline int
+in_to_out_l2_diff(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
+{
+	return iface->l2_len_out - pkt_in_l2_hdr_len(pkt);
+}
+
+/*
+ * Skip the L2 header of the packet, skipping over any VLAN
+ * headers if present. A pointer to the next header is returned.
+ */
+static inline void *
+pkt_out_skip_l2(struct gatekeeper_if *iface, struct ether_hdr *eth_hdr)
+{
+	return ((uint8_t *)eth_hdr) + iface->l2_len_out;
+}
+
+/*
+ * Skip the L2 header of the packet, skipping over any VLAN
+ * headers if present. The EtherType of the next header is returned
+ * (in network order).
+ */
+static inline uint16_t
+pkt_in_skip_l2(struct rte_mbuf *pkt, struct ether_hdr *eth_hdr, void **next_hdr)
+{
+	RTE_VERIFY(next_hdr != NULL);
+
+	pkt->l2_type = RTE_PTYPE_UNKNOWN;
+
+	switch (rte_be_to_cpu_16(eth_hdr->ether_type)) {
+	case ETHER_TYPE_VLAN: {
```

Comment:
Since the constant value in this case (`ETHER_TYPE_VLAN`) is in host ordering, I assumed you meant that I should use the expression `rte_cpu_to_be_16(ETHER_TYPE_VLAN)`, instead of the other way around.

Reply:
You got it right.